### PR TITLE
Block clients from sending server ACLs that lock the local server out.

### LINKED ISSUE #4042: Deny server ACL events in C-S API which would block the server
Its quite easy to accidentally send a server ACL event that blocks everyone, including the current server. I can't think of a time when you would want to block your own server, so lets just deny client requests that try to set ACLs that would block the current server

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #4042

### CODE REVIEW

File: changelog.d/8708.misc [hunk 1]
```diff
@@ -0,0 +1 @@
+Block attempts by clients to sender server ACLs, or redactions of server ACLs, that would result in the local server being blocked from the room.
```
Comment (anoadragon453) [inline changelog.d/8708.misc:1]:
```suggestion
Block attempts by clients to send server ACLs, or redactions of server ACLs, that would result in the local server being blocked from the room.
```

File: synapse/events/validator.py [hunk 2]
```diff
@@ -13,20 +13,25 @@
...
 
 class EventValidator:
-    def validate_new(self, event, config):
+    def validate_new(self, event: EventBase, config: HomeServerConfig):
         """Validates the event has roughly the right format
 
         Args:
-            event (FrozenEvent): The event to validate.
-            config (Config): The homeserver's configuration.
+            event  The event to validate.
```
Comment (anoadragon453) [inline synapse/events/validator.py:34]:
```suggestion
            event: The event to validate.
```

File: tests/handlers/test_message.py [hunk 3]
```diff
@@ -154,3 +154,39 @@ def test_duplicated_txn_id_one_call(self):
...
+        room.register_servlets,
+    ]
+
+    def prepare(self, reactor, clock, hs):
+        self.user_id = self.register_user("tester", "foobar")
+        self.access_token = self.login("tester", "foobar")
+        self.room_id = self.helper.create_room_as(self.user_id, tok=self.access_token)
+
+    def test_allow_server_acl(self):
+        """Test that sending an ACL that blocks everyone but ourselves work.
```
Comment (anoadragon453) [inline tests/handlers/test_message.py:172]:
```suggestion
        """Test that sending an ACL that blocks everyone but ourselves works.
```

File: synapse/handlers/message.py [hunk 4]
```diff
@@ -1159,6 +1159,9 @@ def is_inviter_member_event(e):
                 if original_event.room_id != event.room_id:
                     raise SynapseError(400, "Cannot redact event from a different room")
```
Comment (anoadragon453) [inline synapse/handlers/message.py:1139]: hmm, this should probably be a 403, alas. out of scope for this PR though.

File: tests/handlers/test_message.py [hunk 5]
```diff
@@ -154,3 +154,60 @@ def test_duplicated_txn_id_one_call(self):
...
+        self.helper.send_state(
+            self.room_id,
+            EventTypes.ServerACL,
+            body={},
+            tok=self.access_token,
+            expect_code=400,
+        )
+
+    def test_deny_redact_server_acl(self):
+        """Test that attempting to redact an ACL is blocked
```
Comment (anoadragon453) [inline tests/handlers/test_message.py:195]:
```suggestion
        """Test that attempting to redact an ACL is blocked.
```

**Review by anoadragon453 — APPROVED** [review]
Some wording things, but otherwise looks great functionality-wise!

File: synapse/handlers/message.py (hunk 4 above)
Comment (erkinalp) [reply synapse/handlers/message.py:1139]: No, this is a 400; redacting a room's event using a different room ID is ill-formed, it is not "well-formed but forbidden".

**Pull request merged**